### DISCUSSION TITLE
Is there a way to check if a _thread is still running?

### DISCUSSION THREAD

**Discussion by liudr** [micropython/micropython#10903] [ESP32]
I have an http request handler thread that from time to time receives bad requests that cause it to crash. I've done what I can to handle certain malformed requests but there's always more than what I have currently seen. So I wonder if the main thread can detect whether the http request handler thread has stopped running so it could either restart the thread or reboot. The current doc has no info on _thread which it says is highly experimental. It refers to CPython doc and I don't know if there is a way to detect thread running or stopped. CPython seems to have higher-level threading module that has is_alive() etc but not lower-level _thread functions.
Thanks.

**Comment by peterhinch** [comment]
What do you mean by "crash"? Assuming you mean "throw a Python exception" why can you not trap the exception and terminate the function? If you mean that it blocks (goes into an infinite loop) then you need to subject it to a timeout. It's hard to say how to implement the timeout without knowing exactly where the blocking occurs.

**Comment by liudr** [comment]
Thanks @peterhinch
Yes an exception was thrown so the thread terminated. The exception was an index out of range type, which I later added code to handle. What I am sure will happen is other types of malformed requests that I've not expected will happen and cause another exception that terminates the thread. So I'd like to know if there is a way to detect thread termination.

**Reply by karfas** [reply]
> ... cause another exception that terminates the thread. So I'd like to know if there is a way to detect thread termination.

If unhandled exceptions are the only thing you worry about, **just handle it**. 
What exactly prevents you from wrapping either the entire thread or your request handler in a `try..except` block as @peterhinch suggested ?

**Reply by peterhinch** [reply]
To expand on this, wrap the code in a try-except block so that all exceptions are trapped. When this occurs return a value that the calling code can detect to determine that an error occurred.

**Reply by rkompass** [reply]
That's the way CPythons threading module does it, iirc.
